### ISSUE TITLE
IDE reports "File is not included in module tree, analysis is not available"

### ISSUE THREAD

**Issue by tdyas** [intellij-rust/intellij-rust#6209]
## Environment

* **IntelliJ Rust plugin version:** 0.3.132.3395-202
* **Rust toolchain version:** cargo 1.46.0 (149022b1d 2020-07-17)
* **IDE name and version:** IntelliJ IDEA 2020.2.3 (Ultimate Edition); Build #IU-202.7660.26, built on October 6, 2020
* **Operating system:** macOS 10.15.6

## Problem description

I have a Cargo workspace in a subdirectory (`src/rust`) in a monorepo with `members` set to two crates that are immediate subdirectories of the workspace crate. When trying to edit the Rust files in either crate, the IDE reports `File is not included in module tree, analysis is not available`.

That banner remains even after I invalidate the caches and restart the IDE.

There is a IDE module for Rust defined on the directory for workspace crate.

It had been working when it was just one sub-crate, but adding the second crate caused the issue to re-appear. And the issue persists even after invalidating caches and restarting.

The `idea.log` shows an exception that includes the Rust plugin in its backtrace:

```
2020-10-06 11:21:51,261 [  10688]   INFO - pl.ProjectRootManagerComponent - project roots have changed (trace
_hash = -635222511)
java.lang.Throwable
        at com.intellij.openapi.roots.impl.ProjectRootManagerComponent.synchronizeRoots(ProjectRootManagerCom
ponent.java:306)
        at com.intellij.openapi.roots.impl.ProjectRootManagerComponent.fireRootsChangedEvent(ProjectRootManag
erComponent.java:210)
        at com.intellij.openapi.roots.impl.ProjectRootManagerImpl.fireRootsChanged(ProjectRootManagerImpl.jav
a:428)
        at com.intellij.openapi.roots.impl.ProjectRootManagerImpl.access$200(ProjectRootManagerImpl.java:37)
        at com.intellij.openapi.roots.impl.ProjectRootManagerImpl$BatchSession.rootsChanged(ProjectRootManage
rImpl.java:96)
        at com.intellij.openapi.roots.impl.ProjectRootManagerImpl.makeRootsChange(ProjectRootManagerImpl.java
:371)
        at org.rust.lang.core.macros.MacroExpansionServiceImplInner.stateLoaded(MacroExpansionManager.kt:639)
        at org.rust.lang.core.macros.MacroExpansionManagerImpl$loadState$1$$special$$inlined$invokeLater$1$lambda$1.compute(actions.kt:62)
        at com.intellij.openapi.application.impl.ApplicationImpl.lambda$runWriteAction$16(ApplicationImpl.java:989)
        at com.intellij.openapi.application.impl.ApplicationImpl.runWriteActionWithClass(ApplicationImpl.java:968)
        at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:989)
        at org.rust.lang.core.macros.MacroExpansionManagerImpl$loadState$1$$special$$inlined$invokeLater$1.run(actions.kt:59)
        at com.intellij.openapi.application.TransactionGuardImpl$2.run(TransactionGuardImpl.java:201)
        at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:802)
        at com.intellij.openapi.application.impl.ApplicationImpl.lambda$invokeLater$4(ApplicationImpl.java:322)
        at com.intellij.openapi.application.impl.FlushQueue.doRun(FlushQueue.java:84)
        at com.intellij.openapi.application.impl.FlushQueue.runNextEvent(FlushQueue.java:132)
        at com.intellij.openapi.application.impl.FlushQueue.flushNow(FlushQueue.java:47)
        at com.intellij.openapi.application.impl.FlushQueue$FlushNow.run(FlushQueue.java:188)
        at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
        at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:776)
        at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
        at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
        at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:746)
        at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:971)
        at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:841)
        at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:452)
        at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:744)
        at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$9(IdeEventQueue.java:451)
        at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:802)
        at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:499)
        at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
        at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
        at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
        at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
        at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
        at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

This occurs earlier in the log. Later in the indexing occurring after the invalidate/restart, I see a similar exception but in the Go IDE plugin:

```
2020-10-06 11:29:39,869 [ 479296]   INFO - pl.ProjectRootManagerComponent - project roots have changed (trace
_hash = 174679854)
java.lang.Throwable
        at com.intellij.openapi.roots.impl.ProjectRootManagerComponent.synchronizeRoots(ProjectRootManagerCom
ponent.java:306)
        at com.intellij.openapi.roots.impl.ProjectRootManagerComponent.fireRootsChangedEvent(ProjectRootManag
erComponent.java:210)
        at com.intellij.openapi.roots.impl.ProjectRootManagerImpl.fireRootsChanged(ProjectRootManagerImpl.jav
a:428)
        at com.intellij.openapi.roots.impl.ProjectRootManagerImpl.access$200(ProjectRootManagerImpl.java:37)
        at com.intellij.openapi.roots.impl.ProjectRootManagerImpl$BatchSession.rootsChanged(ProjectRootManage
rImpl.java:96)
        at com.intellij.openapi.roots.impl.ProjectRootManagerImpl.makeRootsChange(ProjectRootManagerImpl.java:371)
        at com.goide.GoLibrariesUtil.lambda$updateLibraries$0(GoLibrariesUtil.java:18)
        at com.intellij.openapi.application.WriteAction.run(WriteAction.java:102)
        at com.goide.GoLibrariesUtil.lambda$updateLibraries$1(GoLibrariesUtil.java:18)
        at com.intellij.openapi.application.TransactionGuardImpl$2.run(TransactionGuardImpl.java:201)
        at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:802)
        at com.intellij.openapi.application.impl.ApplicationImpl.lambda$invokeLater$4(ApplicationImpl.java:322)
        at com.intellij.openapi.application.impl.FlushQueue.doRun(FlushQueue.java:84)
        at com.intellij.openapi.application.impl.FlushQueue.runNextEvent(FlushQueue.java:132)
        at com.intellij.openapi.application.impl.FlushQueue.flushNow(FlushQueue.java:47)
        at com.intellij.openapi.application.impl.FlushQueue$FlushNow.run(FlushQueue.java:188)
        at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
        at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:776)
        at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
        at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
        at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:746)
        at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:971)
        at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:841)
        at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:452)
        at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:744)
        at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$9(IdeEventQueue.java:451)
        at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:802)
        at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:499)
        at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
        at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
        at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
        at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
        at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
        at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

Is there additional information I can collect? Is there a way to enable more logging from the rust plugin or to use a debug version of the plugin with more logging?

**Comment by tdyas** [comment]
```
rustc 1.46.0 (04488afe3 2020-08-24)
binary: rustc
commit-hash: 04488afe34512aa4c33566eb16d8c912a3ae04f9
commit-date: 2020-08-24
host: x86_64-apple-darwin
release: 1.46.0
LLVM version: 10.0
```

**Comment by Undin** [comment]
@tdyas does this workspace is added into plugin project model? You can check it via Cargo tool window
![image](https://user-images.githubusercontent.com/2539310/95251260-f7405c80-0823-11eb-848c-9870aa5b9390.png)

**Comment by tdyas** [comment]
The sub-crates are not listed in that window:

<img width="347" alt="Screen Shot 2020-10-06 at 12 50 17 PM" src="https://user-images.githubusercontent.com/901363/95252994-9a688600-07d2-11eb-8924-4c13dcd35efa.png">

**Comment by tdyas** [comment]
I added the two sub-crates to that window via the + button. I also apparently had a misspecified crate dependency in the `Cargo.toml` for one of the sub-crates that was causing the IDE to not be able to run `cargo metadata`. I fixed that and then clicked the Refresh icon in the Cargo pane. Now everything seems to be fine.

The entry for "rust" now has the two sub-crates showing up underneath it so I removed the two top-level entries that I had added, and still fine.

Maybe some better resilience to error conditions? A misspecified crate dependency shouldn't prevent the plugin from finding projects.

**Comment by vlad20012** [comment]
Do I understand right that you have a cargo workspace with `workspace.members = ["foo", "baz"]` in `Cargo.toml`? In this case, everything _should_ works fine with only the root (workspace) project added in the Cargo pane. So, if you see `File is not included in module tree, analysis is not available`, it means that we have a bug somewhere. But I can't reproduce it: I have such a workspace it works as expected. Maybe your configuration is _a bit_ more complex? E.g. maybe symlinks involved somewhere?

**Comment by Undin** [comment]
@tdyas there some error while cargo project structure loading according to your screenshot https://github.com/intellij-rust/intellij-rust/issues/6209#issuecomment-704515921 (red line under item in Cargo tool window)
You can try to reload project structure (`Refresh Cargo project` action in Cargo tool window) and check `Event Log`

**Comment by tdyas** [comment]
> Maybe your configuration is a bit more complex? E.g. maybe there are symlinks somewhere?

@vlad20012: No symlinks that I'm aware of. One of the crates uses `tonic-build`/`prost` to generate and then use `include!` to bring in Rust code produced by some protobufs. However, the issue showed up when I had a bad version specified for the `clap` crate in the `Cargo.toml` for the other sub-crate. I had just added that other crate to the git repository and the Cargo.toml had the bad version (`3.0` instead of `3.0.0-beta.2`) from the very start.

The source code is in a monorepo where the Rust IntelliJ module is set on the subdirectory `src/rust`. There is another IntelliJ module rooted at the top of the monorepo to handle Python code mostly.

> You can try to reload project structure (Refresh Cargo project action in Cargo tool window) and check Event Log

@Undin: I noticed that after posting initially and as mentioned in https://github.com/intellij-rust/intellij-rust/issues/6209#issuecomment-704519676 ended up clicking that reload button after fixing the version issue with the `clap` dependency. Then everything now works fine.
